Migrate test_assertrewrite.py from testdir to pytester

cmecklenborg opened the pull request:
Partial fix for #7942

Converts test_assertrewrite.py from testdir to pytester

bluetech commented on testing/test_assertrewrite.py:
```diff
@@ -115,18 +115,18 @@ def test_dont_rewrite(self) -> None:
         assert m.body[1].msg is None
 
-    def test_dont_rewrite_plugin(self, testdir):
+    def test_dont_rewrite_plugin(self, pytester: Pytester):
```
It will be great (though not absolutely necessary) if, while adding the `: Pytester` annotation, you also add `-> None` annotations to tests. Without it, mypy (the type checker) still considers the typing to be incomplete. In this case it will be

```suggestion
    def test_dont_rewrite_plugin(self, pytester: Pytester) -> None:
```

bluetech commented on testing/test_assertrewrite.py:
```diff
@@ -135,26 +135,26 @@ def test_rewrites_plugin_as_a_package(self, testdir):
         result.stdout.fnmatch_lines(["*assert 1 == 2*"])
 
-    def test_honors_pep_235(self, testdir, monkeypatch):
+    def test_honors_pep_235(self, pytester, monkeypatch):
```
```suggestion
    def test_honors_pep_235(self, pytester: Pytester, monkeypatch) -> None:
```

bluetech commented on testing/test_assertrewrite.py:
```diff
@@ -135,26 +135,26 @@ def test_rewrites_plugin_as_a_package(self, testdir):
         # note: these modules are named `test_*` to trigger rewriting
-        testdir.tmpdir.join("test_y.py").write("x = 1")
-        xdir = testdir.tmpdir.join("x").ensure_dir()
+        pytester.tmpdir.join("test_y.py").write("x = 1")
```
With `pytester`, `testdir.tmpdir` becomes `pytester.path`. It also becomes a `pathlib.Path`, so the code would need to be a little different. In this case:

```suggestion
        pytester.tmpdir.joinpath("test_y.py").write_text("x = 1", "utf-8")
```

bluetech commented on testing/test_assertrewrite.py:
```diff
@@ -962,26 +962,26 @@ def test():
+        assert pytester.runpytest().ret == 0
 
-    def test_write_pyc(self, testdir: Testdir, tmpdir, monkeypatch) -> None:
+    def test_write_pyc(self, pytester: Testdir, tmpdir, monkeypatch) -> None:
```
Here I suggest also changing `tmpdir` to `tmp_path`, which is the `pathlib.Path` version of `tmpdir`, so it works better with `pytester`.

bluetech commented on testing/test_assertrewrite.py:
```diff
@@ -1062,15 +1062,15 @@ def test_read_pyc(self, tmp_path: Path) -> None:
         assert _read_pyc(source, pyc) is None  # no error
 
-    def test_reload_is_same_and_reloads(self, testdir: Testdir) -> None:
+    def test_reload_is_same_and_reloads(self, pytester: Testdir) -> None:
```
```suggestion
    def test_reload_is_same_and_reloads(self, pytester: Pytester) -> None:
```

bluetech commented on testing/test_assertrewrite.py:
```diff
@@ -1627,11 +1627,11 @@ def test_foo():
 
-        test_foo = Path(testdir.tmpdir) / "src/test_foo.py"
-        bar_init = Path(testdir.tmpdir) / "src/bar/__init__.py"
+        test_foo = Path(pytester.tmpdir) / "src/test_foo.py"
+        bar_init = Path(pytester.tmpdir) / "src/bar/__init__.py"
```
```suggestion
        test_foo = pytester.path / "src/test_foo.py"
        bar_init = pytester.path / "src/bar/__init__.py"
```

bluetech reviewed (changes requested):
Thanks @cmecklenborg, that's a good start on a big file. I left some comments.

cmecklenborg commented:
I moved this to draft status. I addressed PR comments and fixed most of the failing tests, but I still have work to do on 3:
```============================================================================================ short test summary info ============================================================================================
FAILED testing/test_assertrewrite.py::TestAssertionRewrite::test_honors_pep_235 - AssertionError: assert {'errors': 1,...pped': 0, ...} == {'errors': 0,...pped': 0, ...}
FAILED testing/test_assertrewrite.py::TestEarlyRewriteBailout::test_basic - AssertionError: assert None is not None
FAILED testing/test_assertrewrite.py::TestEarlyRewriteBailout::test_pattern_contains_subdirectories - AssertionError: assert None is not None
```

I'll look at this more with fresh eyes tomorrow.

cmecklenborg commented:
@bluetech Do you have any thoughts on why `test_assertrewrite.py::TestEarlyRewriteBailout::test_basic` might be failing on the windows builds?

bluetech commented:
> Do you have any thoughts on why `test_assertrewrite.py::TestEarlyRewriteBailout::test_basic` might be failing on the windows builds?

The issue is this:

```
testing/test_assertrewrite.py:1300: error: Argument 1 to "add" of "set" has incompatible type "str"; expected "local"  [arg-type]
```

It did not show up due to a deficiency in our CI setup, which should be fixed in #7962 for the future.

Explanation of the error: right above this test there is a `hook` fixture which does some funky stuff, one of which is to create mock of a `pytest.Session` called `StubSession`. This uses a `self.initial_paths` attribute, which is a `Set[py.path.local]`. This cannot be changed to `pathlib.Path` since `Session` itself uses `py.path.local` here (for now at least). So to fix the mypy error you'd need to do something like this:

```py
self.initial_paths.add(py.path.local(foobar_path))
```
